src/main/scala/io/iohk/ethereum/mpt/Node.scala

```diff
@@ -4,64 +4,78 @@ import java.util
 
 import akka.util.ByteString
 import io.iohk.ethereum.crypto
-import io.iohk.ethereum.rlp.{encode => encodeRLP}
+import io.iohk.ethereum.rlp.{RLPEncodeable, RLPValue}
 
 /**
   * Trie elements
   */
 sealed abstract class MptNode {
   val cachedHash: Option[Array[Byte]]
   val cachedRlpEncoded: Option[Array[Byte]]
-  import MptNode.MaxEncodedNodeLength
-  import MerklePatriciaTrie._
 
   def withCachedHash(cachedHash: Array[Byte]): MptNode
 
   def withCachedRlpEncoded(cachedEncode: Array[Byte]): MptNode
 
-  lazy val encode: Array[Byte] = cachedRlpEncoded.getOrElse(encodeRLP[MptNode](this))
+  lazy val encode: Array[Byte] = cachedRlpEncoded.getOrElse {
+    parsedRlp.fold(MptTraversals.encodeNode(this))(io.iohk.ethereum.rlp.encode)
+  }
 
   lazy val hash: Array[Byte] = cachedHash.getOrElse(Node.hashFn(encode))
 
-  def capped: ByteString = {
-    val encoded = encode
-    if (encoded.length < MaxEncodedNodeLength) ByteString(encoded) else ByteString(hash)
+  def isNull: Boolean = false
+
+  val parsedRlp: Option[RLPEncodeable]
+
+
+  // Overriding equals is necessery to avoid array comparisons.
+  override def equals(obj: Any): Boolean = {
+    if (!obj.isInstanceOf[MptNode]) {
+      false
+    } else{
+      val compared = obj.asInstanceOf[MptNode]
+      hash sameElements compared.hash
+    }
   }
 
-  def isNull: Boolean = false
+  override def hashCode(): Int = {
+    17 + util.Arrays.hashCode(hash)
```

Comment:
Why 17?

Reply:
Just needed some constant to avoid having `MptNode` having same hashcode as it hash ( which is `Array[Byte]`)
I chose 17 as i remeberd it was used in `Effective Java` - `Store some constant nonzero value, say, 17, in an int variable called result.`

Reply:
maybe add this in a comment, so it would be clear in the future